BAL 3484 - Backend logic for documents tracker data

shanegrouber opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
  - Introduced robust document management capabilities with new APIs for creating, updating, retrieving, and deleting documents.
  - Launched a dedicated document file management system with enhanced file upload handling and metadata tracking.

- **Refactor**
  - Streamlined business processes by removing legacy document fields.
  - Improved file processing by integrating enhanced workflow and storage services.

- **Tests**
  - Expanded automated tests to validate new document management and schema handling functionalities.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

shanegrouber committed:
feat(documents): add documents module and related components

- Introduce DocumentsModule with controller and service
- Implement DocumentTrackerModel for document representation
- Add endpoint to retrieve documents by workflow ID

(Your code is more scattered than a toddler's toys after a playdate)

shanegrouber committed:
Merge branch 'epic/feat/documents-tracker' into BAL-3484

shanegrouber committed:
feat(document): add endpoint to fetch documents by workflow ID

- Implement new route for retrieving documents based on workflow and definition IDs
- Refactor services and modules to accommodate for new functionality

(your code organization is so tangled, it could be mistaken for a Christmas light display)

shanegrouber committed:
feat(documents): implement new document tracking feature

- Add endpoint to retrieve documents based on workflow runtime and definition
- Introduce validation for document input data in the new endpoint

(Your code is so convoluted, it makes a Rubik's Cube look like tic-tac-toe)

shanegrouber committed:
refactor(document): update document service and controller to enhance workflow

- Remove unused DocumentTrackerModel and associated DTOs
- Update getDocumentsByWorkflowId parameters for clarity and consistency
- Refactor document parsing for improved maintainability

(Your function signatures have more parameters than a conference call in 2020)

shanegrouber committed:
chore(workflows-service): update subproject commit and clean up tests

- Update subproject commit reference in data-migrations
- Remove unnecessary console log from document unit test

(Your console logs should come with a warning label: "For debugging only")

shanegrouber committed:
Merge branch 'epic/feat/documents-tracker' into BAL-3484

shanegrouber committed:
Merge branch 'epic/feat/documents-tracker' into BAL-3484

shanegrouber committed:
fix(workflow-service): resolve optional chaining issues in workflow definition

- Remove unnecessary optional chaining for better clarity
- Ensure accurate type handling for workflow definitions

(in just a few lines, your code went from 'nice try' to 'you nailed it' - good job!)

shanegrouber committed:
feat(documents): implement documents tracker items fetching and query

- Add fetch function for documents tracker items
- Create query hooks for retrieving documents with search parameters
- Include hash in URL construction for complete state representation

(Your code organization is so scattered, it looks like a yard sale for functions)

shanegrouber committed:
refactor(document): rename document retrieval functions for clarity

- Change function name from getDocumentsByWorkflowId to getDocumentTrackerByWorkflowId
- Update matching logic for clarity and reduce code duplication

(Your code is like a puzzling riddle—only the compiler knows the answer)

Omri-Levy commented on apps/backoffice-v2/src/domains/documents/hooks/fetchers.ts:
Note the other fetchers files are directly under the domain, in this case /documents/fetchers.ts

Omri-Levy commented on apps/backoffice-v2/src/domains/documents/hooks/query-keys.ts:
```diff
@@ -0,0 +1,15 @@
+import { createQueryKeys } from '@lukemorales/query-key-factory';
+import { fetchDocumentsTrackerItems } from '@/domains/documents/hooks/fetchers';
+
+export const documentsQueryKey = createQueryKeys('documents', {
```
```suggestion
export const documentsQueryKeys = createQueryKeys('documents', {
```

Omri-Levy commented on services/workflows-service/src/document/document.service.ts:
```diff
@@ -264,4 +268,252 @@ export class DocumentService {
+      };
+    }
+
+    const uiSchema = uiDefinition.uiSchema as { elements: Array<Record<string, any>> };
```
Consider runtime validation in the future over a type assertion.

Omri-Levy replied:
Maybe we can combine it with lines 279 through 287

Commit:
feat: document tracker (BAL-3459, BAL-3435) (#3043)

* wip

* interface

* wip

* feat: document tracker

* reorder endpoints

* fallback

* fix

* PR comments + add endpoint

* feature flag

* polish

* invalidate data

* PR comments

Omri-Levy commented on packages/ui/src/components/molecules/AccordionCard/AccordionCard.Item.tsx:
```diff
@@ -50,17 +50,19 @@ export const AccordionCardItem: FunctionComponent<AccordionCardItemProps> = ({
+                  <li
+                    {...liProps}
+                    className={ctw(`flex items-center gap-x-2`, itemClassName, liProps?.className)}
+                    key={index}
```
Let's just ensure text is unique between subitems.
```suggestion
                    key={text}
```

Omri-Levy commented on services/workflows-service/src/document/document.unit.test.ts:
```diff
@@ -0,0 +1,373 @@
+        expect(firstUbo).toBeDefined();
+        expect(secondUbo).toBeDefined();
+
+        if (firstUbo && secondUbo) {
```
Why are the assertions conditional?

shanegrouber committed:
refactor(documents): update query key naming and structure

- Rename documentsQueryKey to documentsQueryKeys for consistency
- Adjust import paths for better organization of document hooks
- Simplify conditional checks in unit tests for better readability

(Your code is cleaner than a whistle, but I still wouldn't recommend using it to clean the house)

Omri-Levy committed:
chore(*): conflicts resolution

Omri-Levy committed:
chore(*): resolved conflicts
